File: AnsatzZurLösungVonFroschProblemen/Platform.cs
using System;
using System.Collections.Generic;

namespace AnsatzZurLösungVonFroschProblemen
{
    public class Platform
    {
        public int indexInArray, deepnessBuffer, rangeOfSight;
        private List<Platform> children;
        private List<int> valueList;

        public Platform(int deepnessBuffer, int indexInArray, List<int> valueList, int rangeOfSight)
        {
            this.deepnessBuffer = deepnessBuffer;
            this.indexInArray = indexInArray;
            this.valueList = valueList;
            this.rangeOfSight = rangeOfSight;
        }

        public List<Platform> hasGoalBeenReached()
        {
            deepnessBuffer--;

            if (deepnessBuffer <= 0)
            {
                if (children == null)
                {
                    children = new List<Platform>();
                    for (int i = indexInArray + 1; i < Math.Min(indexInArray + 1 + rangeOfSight, valueList.Count); i++)
                    {
                        children.Add(new Platform(valueList[i], i, valueList, rangeOfSight));
                        if (children[i - indexInArray - 1].isGoal())
                        {
                            List<Platform> path = new List<Platform>();
                            path.Add(children[i - indexInArray - 1]);
                            path.Add(this);
                            return path;
                        }
                    }
                    return null;
                }

                foreach (Platform p in children)
                {
                    List<Platform> tPath = p.hasGoalBeenReached();
                    if (tPath != null)
                    {
                        tPath.Add(this);
                        return tPath;
                    }
                }
            }

            return null;
        }

        public bool isGoal()
        {
            return indexInArray == valueList.Count - 1;
        }
    }
}

File: scripts/Platform.js
class Platform {
    constructor(depthBuffer, indexInArray, platforms, rangeOfSight, platform_tree, parent) {
        this.depthBuffer = depthBuffer;
        this.indexInArray = indexInArray;
        this.platforms = platforms;
        this.rangeOfSight = rangeOfSight;

        console.log(platform_tree);

        this.children = null;
        this.platform_tree = platform_tree;
        this.node = {
            text: {
                name: this.platforms[this.indexInArray],
            },
        };
        if (parent)
            this.node.parent = parent;

        console.log(this.node);
        this.platform_tree.push(this.node);
    }

    hasGoalBeenReached() {
        this.depthBuffer--;

        if (this.depthBuffer <= 0) {
            if (this.children === null) {
                this.children = [];
                for (let i = this.indexInArray + 1; i < Math.min(this.indexInArray + 1 + this.rangeOfSight, this.platforms.length); i++) {
                    this.children.push(new Platform(Math.abs(Math.abs(this.platforms[this.indexInArray]) - Math.abs(this.platforms[i])), 
                        i, this.platforms, this.rangeOfSight, this.platform_tree, this.node));
                        
                    if (this.children[i - this.indexInArray - 1].isGoal()) {
                        let path = [];
                        path.push(this.children[i - this.indexInArray - 1]);
                        path.push(this);
                        return path;
                    }
                    if (this.children[i - this.indexInArray - 1].depthBuffer == 0) {
                        this.children[i - this.indexInArray - 1].hasGoalBeenReached();
                    }
                }

                return null;
            }

            for (let c of this.children) {
                let tPath = c.hasGoalBeenReached();
                if (tPath !== null) {
                    tPath.push(this);
                    return tPath;
                }
            }
        }

        return null;
    }

    isGoal() {
        return this.indexInArray === this.platforms.length - 1;
    }
}